### DISCUSSION TITLE
Morphologist 2021 pipeline error (sulci_graph is not readable or does not exist)

### DISCUSSION THREAD

**Discussion by Gar1-icm** [brainvisa/brainvisa.github.io#41] [Q&A] [answered]
Hi,
I am a new bainvisa user and I'm working on the Brainvisa 5.0 on a linux system using singularity.
I've tried to run the morphologist pipeline, it works fine through the first steps, than an error blocks the process during the cortical folds recognition, preventing sulci recognicition and graph morphology.
The error message is: _in Morphologist 2021 the parameter left_sulci_graph is not readable or does not exist._
Can someone help me?
I tried using different formats for image entry files (.ima. .nii) and with older pipelines ( 2013, 2015) but it does not run the entire pipeline.
Thanks for your help!
Giacomo

**Comment by denisri** [comment]
Hi,
That's strange. Are all pipeline parameters filled in the main pipeline user interface ?
When it fails, are the "left_graph" and "right_graph" files written ? (can you click the eye icon for them and see the sulcal graphs for unrecognized sulci ?)
If not something has likely failed in a previous step...
Denis

**Reply by Gar1-icm** [reply]
Hi Denis,
thank you very much for your reply.
All the pipeline parameters are checked except normalization. Actually, I've noticed two kinds of errors when running the pipeline: if normalization is checked, the pipeline process stop at 5% and does progress further (without producing any error message). If "normalization" is not checked, the pipeline works fine through the first steps and then blocks at "cortical fold recognition", then stops at 84% without any message error. When trying to open the left_graph file, the  program crashes and is not possible to browse it.

![Screenshot from 2021-04-15 18-00-01](https://user-images.githubusercontent.com/82571025/114901208-dda45100-9e14-11eb-8db2-d126dff078ea.png)

Trying to re-run the pipeline only selecting the remaining steps (cortical folds recognition & sulci recognition), the error "left_sulci_graph is not readable or does not exist" appears. 
I suppose something goes wrong in writing the graph file, but I have no idea about how to fix it...
Any idea? 
thank you very much for your help
Giacomo

**Comment by denisri** [comment]
So I guess the normalization step is the problem. This step is needed to get (at least roughly) a good orientation for the image, several steps later depend on it (mainly the hemispheres split and sulci recognition steps). The normalization algorithm contained in BrainVisa is not the best or the most robust: other software do it well and our focus was not to re-develop algorithms which already work well in other software. So the recommended option for that is to use SPM: you can install SPM standalone version, then configure its path in BrainVisa preferences, and then you get access to a more robust normalization algorithm in the normalization step of BrainVisa.
Unchecking the normalization step is rather intended to images which are already normalized (you can import images already normalized in the MNI space using SPM for instance and process them using Morphologist) - in this case the normalization is already done and you can un-check it. In a "native image" situation, it has to be done.
The last (and oldest) option in Morphologist is to manually click on AC/PC points using the "prepare subject for anatomical pipeline": use the "anatomist" buttons for parameters "anterior_commissue", "posterior_commissure" and "interhemispheric_point", il will open the image in Anatomist viewer, then find the corresponding point, click on it to position the cursor, then click again on the parameter button in the pipeline interface to copy the point position in the pipeline parameter. (see the doc of the process).
Remember you can display the result of each step of the pipeline, so you can check if things go well during execution.

**Comment by Gar1-icm** [comment]
Hi, 
Thank you very much for your answer.
Actually, I un-checked normalization after having performed the “manual” preparation through the “prepare subject for anatomical pipeline" tool. The problem in running the process through the "cortical fold recognition" steps remains despite the manual preparation.
Anyway, I will try using SPM-normalized images to see if it works!
Thank you very much for your kind help!
Giacomo

**Comment by denisri** [comment]
Could you use the "morphologist" program and import the brainvisa database, and post a snapshot of if, so that we may see what has been done correctly and when it went wrong ? Because otherwise it's very difficult to help you "blindly" without seeing any image of the processings.

**Reply by Gar1-icm** [reply]
This is an example of what happens when I load a non normalized volumetric T1 image (directly imported as DICOM through brainvisa and converted into .ima file). I performed manual localization of AC,PC, IC and LHP.
![Screenshot from 2021-04-16 15-26-28](https://user-images.githubusercontent.com/82571025/115032066-2ddceb00-9ec9-11eb-9e4f-02f4370bb823.png)

Then, I run the pipeline un-checking normalization but it abruptly stops at the cortical folds recognition step:
![Screenshot from 2021-04-16 15-31-45](https://user-images.githubusercontent.com/82571025/115032082-31707200-9ec9-11eb-956b-36e7840c3077.png)

At this point, trying to open the files written during the process crashes the program.
When trying to open the graph directly from anatomist, same result:
![Screenshot from 2021-04-16 15-51-16](https://user-images.githubusercontent.com/82571025/115035536-f5d7a700-9ecc-11eb-8b84-00db593f81ac.png)

Actually, this is the case for every file generated during the process (brainvisa crashes and no file can be opened through anatomist).
I can't understand if I do something wrong before running the pipeline, or if this is an error during files writing depending on the software itself.
I hope you can help me. I'm sorry but I am an absolute beginner!
Thank you again for your kind help,
Giacomo

**Comment by denisri** [comment]
OK using manually selected AC, PC etc this seems a correct use of the pipeline.
But I don't see in your snapshots where the pipeline stops abruptly. In the snapshot it seems to be still running in the sulci graph construction step (which can last about a minute, depending on your machine performance), so I don't see anything wrong. Is the pipeline stuck in this exact situation more than 5 minutes ?

Now for the visualization problems, I'm not sure but it might be unrelated, so we have to check that. You say that visualization "crashes the program": you mean that the whole brainvisa exits with a segfault or something similar ? Is there an error message in the terminal (oh yes, please run it from a terminal especially when there are problems so we can see more things) ? What I don't understand is that you provide a snapshot with Anatomist and a nomenclature object loaded: this seems to mean that it is not crashed, but the sulci graph is not loaded.
Can anything else be displayed in 3D ? (for instance try the glxgears program). You show a snapshot of the raw T1 MRI so I guess 3D display is working, however. Is it the same for brain mask or hemispheres split images ?

Did you try using the "morphologist" program as I suggested ?

**Reply by Gar1-icm** [reply]
Dear Denis,
thanks for your reply.
Exactly, the pipeline remains stuck in this exact situation more than an hour, no error is signaled, the process just keep running over and over without going further or producing any new file
This is what the terminal shows when it stops:
_exec: /casa/install/brainvisa/toolboxes/morphologist/startup.py
/host/network/lustre/iss01/cenir/software/irm/spm8: parse directories and
insert items
/host/network/lustre/iss01/cenir/software/irm/spm8: 14 files are stored as 14
DiskItems in 0 seconds
warning: u'/host/network/lustre/iss01/cenir/analyse/irm/users/user/DB_190421' has no uuid
warning: u'/host/network/lustre/iss01/cenir/analyse/irm/users/user/DB_190421' has no uuid
connect to computing resource: icm-cenir-le70
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Exception: <undefined> is not a valid value for data type Unicode
Starting Anatomist.....
config file : /casa/home/.anatomist/config/settings.cfg
PyAnatomist Module present
PythonLauncher::runModules()
0x55ddfa5ef1c0
global modules: /casa/install/share/anatomist-5.0/python_plugins
home   modules: /casa/home/.anatomist/python_plugins
loading module ana_image_math
Anatomist started.
loading module anacontrolmenu
loading module bsa_proba
loading module foldsplit
loading module gradientpalette
loading module histogram
loading module meshsplit
loading module modelGraphs
loading module paletteViewer
loading module palettecontrols
loading module profilewindow
loading module save_resampled
loading module selection
loading module simple_controls
loading module volumepalettes
all python modules loaded
Anatomist started.
Multitexturing present
function glActiveTexture found.
function glClientActiveTexture found.
function glBlendEquation found.
function glTexImage3D found.
function glMultiTexCoord3f found.
function glBindFramebuffer found.
function glBindRenderbuffer found.
function glFramebufferTexture2D found.
function glGenFramebuffers found.
function glGenRenderbuffers found.
function glFramebufferRenderbuffer found.
function glRenderbufferStorage found.
function glCheckFramebufferStatus found.
function glDeleteRenderbuffers found.
function glDeleteFramebuffers found.
Number of texture units: 8
function glUniform1f found.
function glUniform1i found.
function glUniform4fv found.
function glGetUniformLocation found.
function glMultTransposeMatrixf found.
function glAttachShader found.
function glDetachShader found.
function glCompileShader found.
function glCreateProgram found.
function glCreateShader found.
function glDeleteProgram found.
function glDeleteShader found.
function glGetProgramiv found.
function glGetShaderiv found.
function glLinkProgram found.
function glShaderSource found.
function glUseProgram found.
GL_ARB_shadow present
GL_SGIX_shadow extension not available
GL_SGIX_depth_texture extension not available
GL_ARB_depth_texture extension present
GL_ARB_texture_cube_map extension present
GL_EXT_texture_cube_map extension present
Number of texture units: 8
Position : 90.6282, 97.6029, 125, 0
Position : 89.7344, 96.7091, 125, 0
Position : 143.5, 120.167, 125, 0
Position : 94.8333, 106.833, 119.999, 0
Position : 90.1667, 122.833, 119.999, 0
Position : 123.5, 93.5, 119.999, 0
Position : 90.8333, 122.833, 121, 0
Position : 90.4955, 120.998, 121, 0
Position : 90.196, 122.795, 121, 0
Position : 89.8964, 121.597, 121, 0
Position : 135.576, 105.027, 120.997, 0
Position : 91.1252, 66.6385, 86.997, 0
Position : 147.5, 138.833, 113.998, 0_

For the visualization problem, if trying to load the .arg file produced during the pipeline, anatomist freezes while loading and this is what the terminal shows:
_QPixmap::setMask() mask size differs from pixmap size
QPixmap::setMask() mask size differs from pixmap size
memory limit: 126476632064_
If trying to open anatomist and the graph directly from the morphologist window, similarly anatomist freezes  before loading the file, as I showed in the previous message snapshot. This is what the terminal shows:
 __Process **Anatomist Show Fold Graph** started on 2021/04/19 13:32_
 _Process **Anatomist Show Nomenclature** started on 2021/04/19 13:32_
 _Process **Anatomist Show Nomenclature** finished on 2021/04/19 13:32 (0
seconds)_
automatically_labelled: None
manually_labelled: None
memory limit: 140176995123_

And yes, brain masks and hemispeheres split images can be loaded and visualized without problems, it seems that loading .arg file causes the problem.
![Screenshot from 2021-04-19 13-06-58](https://user-images.githubusercontent.com/82571025/115231980-805f1700-a116-11eb-920c-ae8a67b9d01a.png)

I don't know if this information may help you figure out why this is not working.
Thank you very much for your kind help!
Giacomo

**Comment by denisri** [comment]
So I suspect that the sulcal graph is not correctly written, and the reading routine hangs after that. If I understand the same problem (corrupted graph file .arg) could explain all your troubles. Now we have to understand how it could happen (and maybe we should fix something in the reading function because it should not hang anyway, but that's almost another question).
Could you please provide a little snapshot of the hemispheres split and the grey/white segmentation, and then the folds skeleton image, so that I can see if they look OK ? One possibility could be that a wrong hemispheres split would lead to a crappy sulci segmentation.
Or alternately if you can share the data (raw T1 MRI) with me and put it on a temporary server where I can get it to try on my side, that would probably be easier.

**Reply by Gar1-icm** [reply]
Hi Denis,
I sent you a private message with the link to download a sample of the data we've been working on. Let me know if you correctly received it.
I also tried to run the morphologist pipeline from a windows 10 computer with  Brainvisa 5.0 running on a Oracle VM Virtual Box. Strangely, using exactly the same entry data, the pipeline progress a little further producing a left only sulci graph file that can be opened by anatomist. However, the pipeline is interruped by the error during the sulci recognition step and the following error is reported:
_RuntimeError: run through soma workflow failed, see details with soma-workflow-gui_
<img width="289" alt="Cattura" src="https://user-images.githubusercontent.com/82571025/115889237-0dc2a400-a454-11eb-9009-255bd39f1414.PNG">
<img width="363" alt="Cattura2" src="https://user-images.githubusercontent.com/82571025/115896179-6ea1aa80-a45b-11eb-9375-1b54044267e7.PNG">

Un checking the sulci recognition step, the pipeline progresses until the sulci graph morphometry step, when it is interrupeted by the usual message error:
 _in Morphologist 2021 1 the parameter left_sulci_graph is not readable or does not exist_
<img width="470" alt="Cattura1" src="https://user-images.githubusercontent.com/82571025/115896110-5df13480-a45b-11eb-875f-095507468834.PNG">
So I guess the sulci graph is readable but somehow corrupted. 
I don't know if this information let you undestand what happens during pipeline processing.
Thank you very much for your kind help,
Giacomo

**Reply by denisri** [reply]
Hi,
I have received your message and dataset but did not have time to look at it yet. I'll have a look soon.
Denis

**Comment by denisri** [comment, marked as answer]
I have tried processing your data, and had no problem at all. The pipeline ran through the end, and it resulted in correct labelled sulci graphs. I used the internal normalization procedure (AimsMIRegister, which is not the best) without clicking AC/PC points or configuring SPM.
I tried both by importing the DICOM image and using the raw T1 nifti image (not normalized) and both brainvisa-5.0.0 and brainvisa-5.0.1 (so I ran the full pipeline 4 times) and it worked well every time.

**Idea:** Did you check you don't (and didn't) have an "external" problem such as disk full, or disk quota exceeded, or hardware disk IO failure, or anything else ? It could explain many problems (corrupted sulci files, soma-workflow failures...)

**Reply by Gar1-icm** [reply]
Hi Denis,
thank you very much for your kind answer. I am going to look for the "external" problem that prevents a correct use of the pipeline on the Linux system.
For the moment, trying to get around the obstacle, I tried to run the morphologist pipeline from a windows 10 computer with Brainvisa 5.0 running on an Oracle VM Virtual Box, using the same entry data.
In facts, the pipeline can be entirely run selecting “sulci recognition with SPAM”. The .arg files are readable and automatic sulci recognition seem to be correct. However, a red warning message appears during the sulcal recognition process:
_HDF5: infinite loop closing library 
L,T_top,P,P,FD,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E,E_
<img width="301" alt="Cattura4" src="https://user-images.githubusercontent.com/82571025/116235755-87ada280-a75e-11eb-85cc-b0f9faa65be9.PNG">

I already tried to look for its meaning on the support forum but I did not find any correspondence.
If trying to select the default method for sulci recognition (Deep CNN), the following message interrupts the pipeline at this step: _RuntimeError: run through soma workflow failed, see details with soma-workflow-gui_
<img width="289" alt="Cattura" src="https://user-images.githubusercontent.com/82571025/116235831-9c8a3600-a75e-11eb-9083-029be4f07640.PNG">

So my (I hope) last questions for you are:
- What does the “HDF5: infinite loop closing library” error mean? It is something corrupting the pipeline results?
- It is possible that the RuntimeError during Deep CNN sulci recognition process stem from an insufficient amount of memory? Or can causes be different?
- The aim of my work is to calculate length, depth and dorsal-ventral position of the central sulcus and other selected sulci in a group of patients with a neurodevelopmental disorder compared to a control group. In this context, which are the advantages of performing the sulcal recognition with the deep CNN model?

Thank you again for your kind and precious help!
Giacomo

**Comment by denisri** [comment]
> What does the “HDF5: infinite loop closing library” error mean? It is something corrupting the pipeline results?

HDF5 is a library used for reading Minc images format, it's not even used here but sometimes the library issues this message at exit time. I'm not sure if it's a bug in the library itself or in the way we use/initialize it, but it's not harmful.

> It is possible that the RuntimeError during Deep CNN sulci recognition process stem from an insufficient amount of memory? Or can causes be different?

Oh yes it can. The CNN needs about 4-6 Gb of memory and it sometimes fails on machines with 8 Gb of memory or less.
The process is using a newer API than the old brainvisa/axon API, and the error messages are not well integrated, that's why the message only says "use soma_workflow_gui for details". You can either run the `soma_workflow_gui` program and look for the failed job, or, in brainvisa GUI, use the menu "view/workflow execution", which will show a box on the bottom left side of the main window. This box shows computing resources, with by default only your local machine. You can double-click on it to display workflows and jobs and access their ouputs and error logs.
I was thinking of soma-workflow (the distributed calculation tool) itself failing, but you're right, it's probably "just" the job which fails for lack of memory, that's highly probable.

> The aim of my work is to calculate length, depth and dorsal-ventral position of the central sulcus and other selected sulci in a group of patients with a neurodevelopmental disorder compared to a control group. In this context, which are the advantages of performing the sulcal recognition with the deep CNN model?

The CNN model is somewhat more accurate than the older SPAM one, even on the central sulcus (which is the easiest sulcus and has the best results), but all models sometimes make mistakes. So if you have few subjects, and especially if you're interested only in the central sulcus, it's normally a good idea to check visually all results and correct it when it's not perfect (typically there can be little mistakes on branches).
